% !TeX spellcheck = en_US
\documentclass{bioinfo}

\usepackage{hyperref} \usepackage{tikz} \usetikzlibrary{graphs}

\let\proglang=\textsf

\copyrightyear{2015} \pubyear{2015}

\begin{document} \firstpage{1}
	
	\title[RDML]{RDML: an \textbf{R} Package for Working with RDML Format Data}
	\author[Sample \textit{et~al}]{Corresponding Author\,$^{1,*}$, Co-Author\,$^{2}$
		and Co-Author\,$^2$\footnote{to whom correspondence should be addressed}}
	\address{$^{1}$Department of XXXXXXX, Address XXXX etc.\\ $^{2}$Department of
		XXXXXXXX, Address XXXX etc.}
	
	\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}
	
	\editor{Associate Editor: XXXXXXX}
	
	\maketitle
	
	\begin{abstract}
		
		\section{Motivation:} The RDML file format is a recommended element of the
		Minimum Information for Publication of Quantitative Real-Time PCR Experiments
		(MIQE) guidelines. Preserving the setup and output of experiments in
		well-defined scheme, RDML is the universal data storage for qPCR systems.
		Despite the fact that the ability to export data in this format is implemented
		in software products of major real-time PCR instruments, there is no tool
		allowing import and manipulation of RDML data in \textbf{R}.
		
		\section{Results:} We developed \textbf{R} package called RDML, which (i) reads
		fluorescence data obtained during qPCR and melting experiments (ii) extracts
		information about samples (types, targets, concentration etc.) from RDML files
		generated by various real-time PCR instruments; (iii) transforms fluorescence
		data to the appropriate format of the \textit{qpcR} and \textit{chiPCR}
		packages; (iv) generates human readable summary about all samples in experiment;
		(v) creates RDML files from user data. \section{Availability:}
		url{https://cran.r-project.org/web/packages/RDML}. Source code:
		url{https://github.com/kablag/RDML}. \section{Contact:}
		\href{k.blag@yandex.ru}{k.blag@yandex.ru} \end{abstract}
	
	\section{Introduction}
	
	Quantitative real-time PCR (qPCR) based on fluorescence detection is one the
	most popular methods of the molecular biology commonly used across many areas,
	including medicine, food industry  and life sciences. This method is fairly
	simple and at the same time provides accurate, reliable and reproducible
	information\cite{kubista_real-time_2006}. The recently introduced
	high-resolution melting is another actively growing method that relies on
	monitoring fluorescence
	\cite{reed_high-resolution_2007}\cite{wittwer_high-resolution_2009}.
	
	There are about twenty available real-time PCR cycler manufacturers, which in
	turn offers one or more models. Such a variety of models and manufacturers
	creates broader selection of the appropriate models, that differ in many details
	as the number of samples at one run, the number of detection channels, the
	method of work, the price and other parameters. However, such diversification
	comes with a cost. The majority of devices, even produced by a single
	manufacturer, uses different methods of data processing and incompatible data
	storage formats. The discrepancies in data processing hinder comparison of the
	PCR results obtained on different systems. Furthermore, disparate storage
	formats prohibits processing the raw data from one instrument in the another
	system and adding the independent reference methods.
	
	MIQE guidelines were established in 2009 to facilitate the comparison the
	results of qPCR and HRM analysis performed on different instruments
	\cite{bustin_miqe_2009, huggett_2013}. The guidelines suggest using the
	Real-time PCR Data Markup Language (RDML) as the main interchange format for
	qPCR and HRM data \cite{rdml-ninja_2015}. RDML promised to deliver a vendor
	independent and freely available file format, which is based on eXtensible
	Markup Language (XML). Storage and exchange of qPCR data should be
	bidirectional. RDML data standard contains the the raw data acquired by the
	machine, sufficient information to understand the experimental setup (e.g.,
	sample annotation, qPCR protocol, probe and primer sequences) and re-analyse the
	data and interpret the results \cite{lefever_rdml:_2009}. To the best of our
	knowledge RDML is supported by Bio-Rad (CFX 96 and CFX 384), Life Technologies
	(StepOne, ViiA7, QuantStudio) and Roche (LightCycler 96) thermo-cycler systems.
	Third-party software (e.g., primer3plus \cite{untergasser_2007}, QPCR
	\cite{pabinger_2009}, LinRegPCR \cite{ruijter_2014}, \textit{qbase+}
	\cite{hellemans_2007}) are known to process the RDML-format. However, despite
	the openess of the RDML format, there are no open software capable of reading
	and processing this data. Users have to choose between programs from
	manufacturers of PCR devices (which can use only their own generated files) or
	commercial software such as \textit{qbase+} \cite{rdml}. There is only one
	open-source program to visualize, edit and validate RDML files -- RDML-Ninja,
	but it can not analyze fluorescence data \cite{rdml-ninja_2015}. On the other
	hand, there are a number of open source processing solutions for PCR and melting
	curve analysis (e.g., HRM) \cite{roediger_RJ_2013,cousins_2012}, for example,
	implemented in \textbf{R} packages \textit{qpcR}\cite{ritz_qpcr:_2008} and
	\textit{chipPCR} \cite{rodiger2015chippcr}, but they cannot directly read RDML
	files. The RDML package closes this gap and imports RDML data straight to
	\textbf{R} environment.
	
	The availability of an standard is also a prerequisite to many disciplines and
	can accelerate the advance in research. \textbf{R} has advanced internal
	capabilities to record of details about a particular analysis to easily
	recapitulate data from a study \cite{liu_2014}. The common approach for data
	management in \textbf{R} and \textbf{R} graphical user interfaces are native
	formats such as CSV, \textbf{R} workspaces as default data format or objects
	\cite{rodiger_rkward_2012, pabinger_2014, RDCT2014c}. However, a direct exchange
	with other systems was limited. An application of RDML is the distribution of
	data in the unified format. Though there are other approaches available in
	\textbf{R} for reproducible research \cite{Leeper_2014} we think is important to
	bridge the gap to other qPCR as recently reviewed in \cite{pabinger_2014}. However, we 
	argue that the \textit{RDML} package may serve as foundation for other \textbf{R} package
	hosted at Bioconductor \cite{gentleman_2004} and CRAN \cite{RCT} as it was shown
	recently \cite{rodiger2015r}.
	
	In this article we describe our \textbf{R} package \textit{RDML} which allows to
	import qPCR data from RDML~v1.1 or v1.2 format files and transform it to the
	human readable format appropriate to the \textit{qpcR} and \textit{chipPCR}
	packages. \textit{RDML} package is open-source, free to use,
	platform-independent, and thus can be modified for specific tasks or other
	\textbf{R} packages, or even integrated to work-flows with other programming
	languages. Despite though many qPCR instruments and software solutions are able
	work with the RDML format, there is a need make RDML available for the ever
	increasing used \textbf{R} environment.
	
	\section{Implementation}
	
	The most recent stable \textit{RDML} package is available at CRAN
	(\url{http://cran.r-project.org/web/packages/RDML/index.html}) and the
	development version is hosted at GitHub (\url{https://github.com/kablag/RDML})
	with facilities to report bugs or request features. All package development
	followed the guidelines as descibed elsewhere \cite{RDCT2014a}. \textit{RDML}
	provides several \textbf{R6} classes, which corresponds to RDML~v1.2 format types.
	\textbf{R} is language with dynamic typing, which is helpful while scripting but
	can lead to problematic debugging of more complicated workflows. \emph{R6} classes
	provide type-safe interfaces to set data without access to inner structure of
	objects thus all inputed data can be validated. This option is very useful when
	creating packages to work as intermediate level for other packages (e.g., such
	approach does not allow set \textbf{character} in place of \textbf{integer}
	Supplementary Section S3). Furthermore, the inheritance of \textbf{R6} objects 
	unifies the structure of the package and streamlines extending its capabilities 
	(the whole package is written around a single base class).
	
	The RDML file format, coordinated by the RDML consortium (www.rdml.org/), is still 
	in development. The \textit{RDML} package follows the changes and supports the 
	current version RDML~1.2. The future improvements of the RDML format will be also 
	incorpotated into the package.
	
	Central functions of the \textit{RDML} package encompass the read-in of RDML
	data file and summary functions for RDML objects. The capabilities of packages range 
	from advanced statistical analysis to simple yet crucial extraction of the fluorescence 
	data. Manipulations with the RDML data can be done by a major class called RDML
	through
	its public methods. These methods include: \begin{itemize} \item
		\textbf{\$new()} -- creates new RDML object. Empty or from specified RDML file
		(Supplementary Section S1); \item \textbf{\$AsDendrogram()} -- represents
		structure of RDML object as dendrogram to observe it. (Supplementary Section
		S1); \item \textbf{\$AsTable()} -- represents data contained in RDML object
		(except fluorescent data) as \textit{data.frame} (Supplementary Section S4);
		\item \textbf{\$GetFData()} -- gets fluorescent data (Supplementary Section S5);
		\item \textbf{\$SetFData()} -- sets fluorescent data to RDML object
		(Supplementary Section S6); \item \textbf{\$AsXML()} -- saves RDML object as
		RDML~v1.2 file (Supplementary Section S7). \end{itemize}
	
	Beside working with existing RDML files, the package can create new RDML objects from
	the data obtained from systems that do not support RDML format. To create eligible RDML 
	object, one has to provide fluorescence data and minimal description of experiment
	(Supplementary Section S10).
	
	The unique advantage of our package is ability to merge several RDML files
	(experiments) into one with \textbf{MergeRDMLs()} function and process them like
	unit (Supplementary Section S7). For example, you can combine results of two
	runs that have sample from one experiment or add calibration samples from one
	run to another.
	
	We join the claim that data sets are an essential element of \textbf{R} packages 
	for reproducible research \cite{gentleman_2004,hofmann_2013,Leeper_2014}. 
	As of \textit{RDML} v.~0.9-1 we provide data sets from the Bio-Rad CFX 96, Roche 
	LightCycler 96 and Applied Biosystems StepOne.
	
	\section{Discussions and conclusions}
	
	The \textit{RDML} package for \textbf{R} imports and manipulates data from RDML~v1.1 and 
	v1.2 format files. It can also create RDML~v1.2 files from data provided by an user. 
	This package can be used as a part of the qPCR or HRM data processing workflows or 
	for experiments overview.
	
	\paragraph{Funding\textcolon}  Part of this work was funded by the Russian
	Ministry of Education and Science (project No. RFMEFI62114X0003), with usage of
	scientific equipment of Center for collective use ’Biotechnology’ at All-Russia
	Research Institute of Agricultural Biotechnology.
	
	%\bibliographystyle{natbib}
	%\bibliographystyle{achemnat}
	%\bibliographystyle{plainnat}
	%\bibliographystyle{abbrv}
	%\bibliographystyle{bioinformatics}
	%
	\bibliographystyle{plain} %
	\bibliography{RDML}
	
\end{document}
